engine/consensus/provider/engine.go

```diff
@@ -168,3 +168,39 @@ func (e *Engine) onBlockProposal(originID flow.Identifier, proposal *messages.Bl
 
 	return nil
 }
+
+// recipientList returns a list of nodes which we should include when publishing
+// blocks. We include:
+// * all staked non-consensus nodes in the current epoch
+// * all non-consensus nodes registered for the next epoch, if known
+func (e *Engine) recipientList() (flow.IdentityList, error) {
+
+	final := e.state.Final()
+
+	// we always include non-consensus nodes from the current epoch
+	currentEpochRecipients, err := final.Identities(filter.Not(filter.HasRole(flow.RoleConsensus)))
+	if err != nil {
+		return nil, fmt.Errorf("could not get current epoch recipients: %w", err)
+	}
+
+	phase, err := final.Phase()
+	if err != nil {
+		return nil, fmt.Errorf("could not get current epoch phase: %w", err)
+	}
+
+	// we don't know the nodes for the next epoch yet
+	if phase == flow.EpochPhaseStaking {
+		return currentEpochRecipients, nil
+	}
+
+	// once the next epoch has been set up, include those nodes
+	nextEpochIdentities, err := final.Epochs().Next().InitialIdentities()
+	if err != nil {
+		return nil, fmt.Errorf("could not get next epoch recipients: %w", err)
+	}
+
+	// filter to only include non-consensus nodes
+	nextEpochRecipients := nextEpochIdentities.Filter(filter.Not(filter.HasRole(flow.RoleConsensus)))
+
+	return currentEpochRecipients.Union(nextEpochRecipients), nil
```

Comment:
Is it worth to cache the recipientList, otherwise it always have to union when sending a block.

Reply:
I'd rather keep it simple. Even for thousands of nodes, this will be very inexpensive. Plus, when we have slashing, the cache could be potentially be invalidated every block.

Reply:
Right. we are proposing 1 block per second, so should be OK